File: python/spiral-matrix/spiral_matrix.py
"""
Module providing a function to generate a spiral matrix.
"""

def spiral_matrix(size):
    """
    Given the size, return a square matrix of numbers in clockwise spiral order.
    """
    matrix = []
    if size > 0:
        matrix = [[0 for col in range(size)] for row in range(size)]

        top_row = 0
        bottom_row = size - 1
        left_col = 0
        right_col = size - 1
        fill_number = 1

        while fill_number <= size * size:
            for column in range(left_col, right_col + 1):
                matrix[top_row][column] = fill_number
                fill_number += 1

            top_row += 1

            for row in range(top_row, bottom_row + 1):
                matrix[row][right_col] = fill_number
                fill_number += 1

            right_col -= 1

            for column in range(right_col, left_col - 1, -1):
                matrix[bottom_row][column] = fill_number
                fill_number += 1

            bottom_row -= 1

            for row in range(bottom_row, top_row - 1, -1):
                matrix[row][left_col] = fill_number
                fill_number += 1

            left_col += 1

    return matrix

File: cpp/spiral-matrix/spiral_matrix.cpp
#include "spiral_matrix.h"

namespace spiral_matrix {
    std::vector<std::vector<uint32_t>> spiral_matrix(int spiral_size)
    {
        std::vector<std::vector<uint32_t>> result(spiral_size, std::vector<uint32_t>(spiral_size, 0));

        int top_row = 0;
        int bottom_row = spiral_size - 1;
        int left_col = 0;
        int right_col = spiral_size - 1;
        uint32_t fill_number = 1;

        while(fill_number <= static_cast<uint32_t>(spiral_size * spiral_size)) {
            for(int column = left_col; column <= right_col; column++) {
                result[top_row][column] = fill_number++;
            }

            top_row += 1;

            for(int row = top_row; row <= bottom_row; row++) {
                result[row][right_col] = fill_number++;
            }

            right_col -= 1;

            for(int column = right_col; column >= left_col; column--) {
                result[bottom_row][column] = fill_number++;
            }

            bottom_row -= 1;
            
            for(int row = bottom_row; row >= top_row; row--) {
                result[row][left_col] = fill_number++;
            }

            left_col += 1;
        }

        return result;
    }
}  // namespace spiral_matrix
